### PR TITLE
Timberland final v3

### PR DESCRIPTION
**Pull request by johnmeneghini**
These are the patches I'm using in the rh-linux-poc.

Note the most recent patch, which enables everything to work with @stuarthayes latest nvme-cli changes.

## Checklist
- [ x] I have tested it locally
- [ ] I have reviewed and updated any documentation if relevant
- [ x] I am providing new code and test(s) for it

Fixes #

### CODE REVIEW

**Review by mwilck — CHANGES_REQUESTED** [review]
This must be merged with #5. Also, the `connect-all` logic must be discussed.

With the merge of the nvme-cli code, the situation is different than before. We now have the `connect-all` command that supports both NBFT and ordinary discovery. Thus we could simplify the dracut logic substantially by simply running `nvme connect-all`, without distinguishing the various cases (NBFT present, `discovery.conf` present, `config.json` present, etc.) as `connect-all` carries out the same tests again.

This comes down to the discussion I had previously with @hreinecke, who argued that we should simply use `connect-all`, always, whereas I think that for the purpose of activating the root file system, it makes sense to limit ourselves to those subsystems/namespaces that are actually necessary. At least, it should be **possible** for the admin to enforce such a limitation (e.g. if either the NBFT or the static `discovery.conf` contains broken data, which might lead to timeouts or even a wrong root file system mounted).

We have `rd.nvmf.nonbft` already and we need to keep supporting that.
IMO that would imply that we must call `connect-all --no-nbft` if this parameter is given.

Likewise, IMO we would need `rd.nvmf.nbft` or `rd.nbft.nbftonly` if we want to allow the admin to force use of **only** NBFT provided subsystems. This would require calling `nvme connect-all --nbft`.

Now we have the problem that all versions of `nvme` except the latest will error out with these command line options. So we need to  implement a fallback.

Wrt static discovery (`discovery.conf`, `config.json`), we must also keep in mind that TCP and RDMA connection attempts to non-reachable network destinations might fail or worse, time out, unless dracut has been configured to bring up the network. Only NBFT configuration basically ensures that with every configured subsystems comes a matching HFI entry, and that these subsystems will probably be reachable.

AFAICS, we have two options now:

1. Just use `connect-all`, as suggested by @hreinecke, and don't bother about using NBFT or local configuration. That would simplify the dracut code substantially, and it would (sort of) minimize the risk that the root FS isn't found we donn't connect to the subsystem that contains it. It suffers from the issues mentioned above though.

2. Fix the existing logic to handle the various cases mentioned above.

Not surprisingly, I vote for 2. I'd like to base our upstream dracut PR on 2.), and have a discussion about the `connect-all` logic in the upstream PR. If upstream comes to the conclusion that the complexity of differentiating the various cases is unnecessary, we can still simplify the PR.

File: modules.d/95nvmf/nvmf-autoconnect.sh [hunk 1]
```diff
@@ -17,7 +17,7 @@ if [ -e /tmp/nvmf-fc-auto ] && [ "$NVMF_HOSTNQN_OK" ] \
 fi
 if [ -e /tmp/valid_nbft_entry_found ]; then
     # prio 2: NBFT
-    /usr/sbin/nvme connect-nbft
+    /usr/sbin/nvme connect-all
```
[the remark this answers is not in the document]
Comment (johnmeneghini) [reply modules.d/95nvmf/nvmf-autoconnect.sh:20]: No, the upstream change code does not require this.  This patch is from @charles-rose

[the remark this answers is not in the document]
Comment (johnmeneghini) [reply modules.d/95nvmf/nvmf-autoconnect.sh:20]: fixed

File: modules.d/95nvmf/nvmf-autoconnect.sh [hunk 2]
```diff
@@ -17,7 +17,7 @@ if [ -e /tmp/nvmf-fc-auto ] && [ "$NVMF_HOSTNQN_OK" ] \
...
     [ "$1" = timeout ] || exit 0
 fi
 if [ -f /etc/nvme/discovery.conf ] && [ $NVMF_HOSTNQN_OK ]; then
```
[the remark this answers is not in the document]
Comment (johnmeneghini) [reply modules.d/95nvmf/nvmf-autoconnect.sh:23]: fixed

**Comment by johnmeneghini** [comment]
> This must be merged with #5. Also, the `connect-all` logic must be discussed.
> 
> AFAICS, we have two options now:
> 
> 1. Just use `connect-all`, as suggested by @hreinecke, and don't bother about using NBFT or local configuration. That would simplify the dracut code substantially, and it would (sort of) minimize the risk that the root FS isn't found we donn't connect to the subsystem that contains it. It suffers from the issues mentioned above though.
> 2. Fix the existing logic to handle the various cases mentioned above.
> 
> Not surprisingly, I vote for 2. I'd like to base our upstream dracut PR on 2.), and have a discussion about the `connect-all` logic in the upstream PR. If upstream comes to the conclusion that the complexity of differentiating the various cases is unnecessary, we can still simplify the PR.

OK. let's go with option 2.  I will l make the appropriate changes and then rebase both  `timberland_final_v3` and `timberland_final_v3-mwilck` onto the the `timberland_final`

File: modules.d/95nvmf/nvmf-autoconnect.sh (hunk 1 above)
[the remark this answers is not in the document]
Comment (johnmeneghini) [reply modules.d/95nvmf/nvmf-autoconnect.sh]: I have fixed this problem and rebased this branch onto `timberland_final`

File: modules.d/95nvmf/nvmf-autoconnect.sh (hunk 2 above)
[the remark this answers is not in the document]
Comment (johnmeneghini) [reply modules.d/95nvmf/nvmf-autoconnect.sh]: I have fixed this problem and rebased this branch onto `timberland_final`

**Comment by johnmeneghini** [comment]
This pull request has merged #7 and is now ready to be merged into timberland _final.

**Review by mwilck — APPROVED** [review]
My suggestions have been merged, thanks!

**Comment by mwilck** [comment]
@johnmeneghini: I will rebase-merge this on the command line, because I strongly dislike github's rebase-and-merge behavior.

**Review by mwilck — CHANGES_REQUESTED** [review]
@johnmeneghini,

something went wrong here. In your merge-and-squash operation, you lost part of my commit 48508e5, and entirely removed 83a5260.

Please double-check.

**Comment by mwilck** [comment]
Something went wrong merging #7 into this PR. I can't merge this into the timberland_final branch just yet.

**Comment by johnmeneghini** [comment]
> @johnmeneghini,
> 
> something went wrong here. In your merge-and-squash operation, you lost part of my commit [48508e5](https://github.com/timberland-sig/dracut/commit/48508e55b46eb5a25b835dee2976c1d300a944ac), and entirely removed [83a5260](https://github.com/timberland-sig/dracut/commit/83a5260f685fa135df2e96c3b4157287f05b46f6).
> 
> Please double-check.

OK I will look at this.

**Comment by johnmeneghini** [comment]
I've merged your pull request #8 into this branch and now the Commisery test is failing on your patch.  I have no idea why this is happening because all of the commit headers look correct to me.  The last time I encountered this error it was because one of the patches had a "signed-off-by" label.

I'm still not sure how I lost your change in the first place.  IIRC it was probably because of a problem like this. I hate software that thinks it knows what I want... from now I am going to merge all of my changes with a forced push.  That way I know the SHA will be added to the branch...

Please just squash this branch and merge it.  Note that you may need to use your branch from #8 for squashing and merging because Github has not added this patch to the branch for #6 yet.

**Comment by mwilck** [comment]
Force-pushed this branch to 0aba896 just to get rid of the merge commit that commisery doesn't like.

**Comment by johnmeneghini** [comment]
> Force-pushed this branch to [0aba896](https://github.com/timberland-sig/dracut/commit/0aba8963b54a15595b3e06356e085fa367a2a0f5) just to get rid of the merge commit that commisery doesn't like.

OK. if you are happy with this branch you should go a head and merge it.

**Comment by mwilck** [comment]
Rebased and force-pushed again. No code changes. I just squashed the entire NBFT related code into a single commit, and rewrote the commit description.

Last (trivial) step will be rebase to current upstream HEAD.

@johnmeneghini, @charles-rose, please have a final look.

**Review by johnmeneghini — COMMENTED** [review]
This all looks good Martin.  Thanks for updating this.

**Comment by johnmeneghini** [comment]
Why does this branch have merge conflicts?

I think you need to rebase everything here to the tip of the master branch.

File: modules.d/95nvmf/nvmf-autoconnect.sh [hunk 3]
```diff
@@ -1,5 +1,54 @@
-#!/bin/bash
+#!/bin/sh
```
[the remark this answers is not in the document]
Comment (mwilck) [reply modules.d/95nvmf/nvmf-autoconnect.sh:1]: It should have been one from day 1. Scripts that run during boot should be POSIX shell, not bash.

**Comment by mwilck** [comment]
> Why does this branch have merge conflicts?

They are merge conflicts with the current `timberland_final` branch (where the old nvme-cli command line syntax was used), not with upstream. I just tried on the command line, and there are also merge conflicts between `timberland_final` and your `timberland_final_v4-johnm` branch, the same as here. They could be trivially resolved, but eventually we're going to force-push anyway, so I don't think this is an issue.

> I think you need to rebase everything here to the tip of the master branch.

With "master" you mean "upstream master"? That wouldn't make a difference, the merge conflicts with `timberland_final` would remain, as they're between the old and new timberland code.

**Comment by johnmeneghini** [comment]
On 5/3/23 05:06, Martin Wilck wrote:
> With "master" you mean "upstream master"? That wouldn't make a difference, the merge conflicts with |timberland_final| would 
> remain, as they're between the old and new timberland code.

Yes, I mean I think we should rebase the upstream pull request `timberland_final` to the HEAD of the upstream master branch.

**Comment by mwilck** [comment]
@johnmeneghini,  @charles-rose: You still didn't say anything about my rebase operation and the rewritten commit message.

**Review by mwilck — APPROVED** [review]
I was still listed as requesting changes… I obviously approve this PR now.

**Comment by mwilck** [comment]
I'm going to "merge" this now by force-pushing this code to `timberland_final`.

**Comment by mwilck** [comment]
With all changes from this branch now "merged" into `timberland_final`, we're good to close here.

**Pull request merged**